kv/dist_sender_test.go

```diff
@@ -752,3 +752,99 @@ func TestGetNodeDescriptor(t *testing.T) {
 		return util.Errorf("wanted NodeID 5, got %v", desc)
 	})
 }
+
+// TestSpanRangeScanWhenMerging simulate a range merge operation happend between get range desc
+// and get next range desc in getDescriptors function. the second sub scan use next range's start key
+// as start key of scan, this cause the the whole scan get duplicate key-values.
+func TestSpanRangeScanWhenMerging(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	g, s := makeTestGossip(t)
+	defer s()
+
+	// Asume we have two range [a-b) [b-KeyMax)
+	merged := false
+
+	var firstRange = proto.RangeDescriptor{
+		RaftID:   1,
+		StartKey: proto.Key("a"),
+		EndKey:   proto.Key("b"),
+		Replicas: []proto.Replica{
+			{
+				NodeID:  1,
+				StoreID: 1,
+			},
+		},
+	}
+
+	// Two ranges will be merged to one range
+	var mergedRange = proto.RangeDescriptor{
+		RaftID:   1,
+		StartKey: proto.Key("a"),
+		EndKey:   proto.KeyMax,
+		Replicas: []proto.Replica{
+			{
+				NodeID:  1,
+				StoreID: 1,
+			},
+		},
+	}
+	// Asume we have two key-values: a-1  c-2
+	var testFn rpcSendFn = func(_ rpc.Options, method string, addrs []net.Addr, getArgs func(addr net.Addr) interface{}, getReply func() interface{}, _ *rpc.Context) ([]interface{}, error) {
+		header := getArgs(testAddress).(proto.Request).Header()
+		reply := getReply().(*proto.ScanResponse)
+		if method == "Node.Scan" && header.Key.Equal(proto.Key("a")) && header.EndKey.Equal(proto.Key("b")) {
+
+			reply.Rows = []proto.KeyValue{
```

Comment:
the comparisons should simplify if you do something like
```
[]results = []proto.KeyValue{}

if header.Key.Less(aKV.Key.Next()) && aKV.Key.Less(header.EndKey) {
   results = append(results, aKV)
}
...
```

you should only need two `if`s then.